glib/src/value.rs

```diff
@@ -56,31 +51,264 @@ use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
 
-/// An error returned from the [`get`](struct.Value.html#method.get)
-/// or [`get_some`](struct.Value.html#method.get_some) functions on a [`Value`](struct.Value.html)
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct GetError {
-    pub actual: Type,
-    pub requested: Type,
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling optional types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is optional.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeOptional:
+    ValueType
+    + ToValueOptional
+    + for<'r> FromValue<'r, Error = ValueTypeMismatchOrNoneError<Self>>
+    + StaticType
+{
+}
+
+impl<'r, T: ValueTypeOptional + StaticType> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = ValueTypeMismatchOrNoneError<T>>,
+{
+    type Type = T::Type;
+}
+
+/// Trait for errors returned from `Value::get`.
+pub unsafe trait ValueTypeChecker: std::error::Error + Sized {
+    fn check(value: &Value) -> Result<(), Self>;
```

Comment:
Will become `fn check(value: &Value, type_: Type) -> Result<(), Self>` for the same reason

Reply:
Or maybe rather `fn check<T: StaticType>(value: &Value) -> Result<(), Self>`, not sure...

Reply:
I think `fn check<T: StaticType>(value: &Value) -> Result<(), Self>` is better if that causes no issues.

Reply:
Thanks, I couldn't decide between the two :) Neither causes problems from what I can see so far.

Reply:
That all doesn't fit very well either, no matter which of the two. Hm!

Reply:
Basically the error types need to be generic over the type or otherwise they can't actually implement any checking whether the value is compatible with it.